src/nvim/lua/vim.lua

```diff
@@ -457,6 +457,93 @@ function vim.region(bufnr, pos1, pos2, regtype, inclusive)
   return region
 end
 
+--- Get the region between two marks and the start and end positions for the region
+---
+--@param mark1 Name of mark starting the region
+--@param mark2 Name of mark ending the region
+--@param options Table containing the adjustment function, register type and selection mode
+--@return region region between the two marks, as returned by |vim.region|
+--@return start (row,col) tuple denoting the start of the region
+--@return finish (row,col) tuple denoting the end of the region
+function vim.get_marked_region(mark1, mark2, options)
```

Comment:
This can use anything that can be passed to `getpos()`, which includes but is not limited to marks. Calling it `get_marked_region` is misleading and (IMO) a bit confusing.

Can we modify `vim.region` to accept the special characters that `line()` and friends take, in addition to `(row, col)` tuples:

```
.            the cursor position
$            the last line in the current buffer
'x           position of mark x (if the mark is not set, 0 is
             returned)
w0           first line visible in current window (one if the
             display isn't updated, e.g. in silent Ex mode)
w$           last line visible in current window (this is one
             less than "w0" if no lines are visible)
v            In Visual mode: the start of the Visual area (the
             cursor is the end).  When not in Visual mode
             returns the cursor position.  Differs from |'<| in
             that it's updated right away.
```

If so, then we wouldn't need this new function at all and we can just use `vim.region()`.

Reply:
Should be a straightforward refactor by moving the code (specific to `'[` and `']`) from `vim.highlight.on_yank` up to `vim.region` and just passing those marks through.

(Handling multibyte characters is a separate and more general issue that I'd recommend tackling as a follow-up PR.)

Reply:
Done in https://github.com/neovim/neovim/pull/22635